cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java

```diff
@@ -138,6 +142,28 @@ public void testPipelineWithAllActions() throws Exception {
     Assert.assertEquals(action1Value, MockAction.readOutput(actionTableDS, action1RowKey, action1ColumnKey));
     Assert.assertEquals(action2Value, MockAction.readOutput(actionTableDS, action2RowKey, action2ColumnKey));
     Assert.assertEquals(action3Value, MockAction.readOutput(actionTableDS, action3RowKey, action3ColumnKey));
+
+    List<RunRecord> history = workflowManager.getHistory(ProgramRunStatus.COMPLETED);
+    Assert.assertEquals(1, history.size());
+    String runId = history.get(0).getPid();
+
+    WorkflowTokenDetail tokenDetail = workflowManager.getToken(runId, WorkflowToken.Scope.USER,
+                                                               action1RowKey + action1ColumnKey);
+
+    validateToken(tokenDetail, action1RowKey + action1ColumnKey, action1Value);
+
+    tokenDetail = workflowManager.getToken(runId, WorkflowToken.Scope.USER, action2RowKey + action2ColumnKey);
+    validateToken(tokenDetail, action2RowKey + action2ColumnKey, action2Value);
+
+    tokenDetail = workflowManager.getToken(runId, WorkflowToken.Scope.USER, action3RowKey + action3ColumnKey);
+    validateToken(tokenDetail, action3RowKey + action3ColumnKey, action3Value);
+  }
+
+  private void validateToken(WorkflowTokenDetail tokenDetail, String key, String value) {
+    List<WorkflowTokenDetail.NodeValueDetail> nodeValueDetails = tokenDetail.getTokenData().get(key);
```

Comment:
i'm not sure if this key is node name or the actual key, as the same key is used at both places.